Step2 문자열 계산기 코드리뷰 요청드립니다.

dongha-byun opened the pull request:
안녕하세요!
우아한 테크캠프 Pro 5기 교육생 변동하 입니다.

2단계 과제 문자열 계산기 코드리뷰 요청드립니다.

감사합니다.

dongha-byun committed:
Test: 문자열 계산시 테스트코드 작성

dongha-byun committed:
Feat: 문자열 계산기 구현코드 작성
- 빈 문자열은 0으로 반환
- null 문자열은 0으로 반환
- 쉼표(,) 기준으로 덧셈 연산 처리

dongha-byun committed:
Test: 문자열 덧셈 계산기 테스트코드 작성

dongha-byun committed:
Feat: 문자열 계산기 기능 구현

dongha-byun committed:
Refactor: 문자열 계산기 구현코드 리펙터링
- 구현코드 리펙터링에 따라 테스트코드도 같이 변경

chae-yh commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class StringCalculator {
```
해당 기능에 대한 요구사항을 정리해보면 어떨까요...?
그리고, 해당 클래스가 자르고, 더하는 등 너무 많은 역할을 하고 있는 것 같아요. 조금 더 역할 분리를 해보면 어떨까요...?

dongha-byun replied:
네 알겠습니다! 클래스를 좀 더 작게 분리하고 그에 따라 테스트코드도 같이 수정해보겠습니다 ㅎㅎ

chae-yh commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    private static int addByDelimiter(String input, String delimiter) {
+        AtomicInteger result = new AtomicInteger();
```
이 부분은 궁금해서 여쭤봅니다. 혹시 AtomicInteger 을 사용하신 이유가 있을까요...?

dongha-byun replied:
사실 이 부분은 뭘 알고 쓴건 아니었고, IDE로 리팩토링하다보니 자동으로 바뀐부분입니다...ㅠ

chae-yh commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,60 @@
+
+    public static int addOperation(String input) {
+        if(validateNullOrEmpty(input)){
+            return 0;
```
이런 매직 넘버는 상수로 extract 해보는 것은 어떨까요...?

dongha-byun replied:
오 바꿀까 말까 고민했는데 역시 바꾸는게 나았겠네요! 수정해보겠습니다!

chae-yh commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,60 @@
+            return 0;
+        }
+
+        Matcher matcher = Pattern.compile(CUSTOM_CALCULATOR_REGEX).matcher(input);
```
Pattern.compile 은 비용이 많이 듭니다. 이 부분을 더할 때마다 해주지 말고 미리 compile을 해두면 어떨까요...?

dongha-byun replied:
생각 못한 부분이었어요...! 말씀해주신대로 수정해보겠습니다!

chae-yh commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,63 @@
+    @Test
+    @DisplayName("빈 문자열은 0으로 반환한다.")
+    void emptyStringIsZeroTest() {
+        assertThat(StringCalculator.isEmpty("")).isTrue();
```
이 테스트는 "빈 문자열은 0으로 반환한다"에 대한 테스트는 아닌 것 같아요. 
empty 여부를 판단할 수 있다는 테스트로 분리해볼 수 있지 않을까요...?

dongha-byun replied:
앗.. 코드 리펙토링하다가 저 이름은 놓쳤나봐요 ㅠㅠ 다시 한 번 확인해서 적절한 이름으로 고쳐보겠습니다!

chae-yh commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,63 @@
+    @DisplayName("빈 문자열은 0으로 반환한다.")
+    void emptyStringIsZeroTest() {
+        assertThat(StringCalculator.isEmpty("")).isTrue();
+        assertThat(StringCalculator.addOperation("")).isEqualTo(0);
```
```suggestion
        assertThat(StringCalculator.addOperation("")).isZero();
```
로 써볼 수도 있을 것 같습니다 ㅎㅎ

dongha-byun replied:
헐 코드 까지 작성해주시다니 감사드립니다 ㅠㅜ

chae-yh commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,63 @@
+    @ParameterizedTest
+    @CsvSource(value={"1,2:3", "1,2,3:6", "2,4,:6", "1:1"}, delimiterString = ":")
+    @DisplayName("쉼표(,)를 구분자로 취급하여 덧셈한다.")
+    void plusOperationByComma(String input, int result){
```
@ParameterizedTest를 잘 사용해주셨네요 👍

dongha-byun replied:
감사합니다 ^0^

chae-yh commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,60 @@
+
+    public static void validateNumber(String input) {
+        if(!Pattern.matches(ONLY_NUMBER_REGEX, input)){
+            throw new RuntimeException("숫자 이외의 문자가 포함되어 있습니다.");
```
RuntimeException은 너무 범위가 큰 exception인 것 같아요. IllegalArgumentException 같은 것은 어떨까요...?

dongha-byun replied:
네! 그게 좋을것 같습니다. 말씀해주신대로 수정해보겠습니다!

chae-yh reviewed (changes requested):
동하님 안녕하세요 : ) 두 번째 pr을 벌써 주셨군요 👍 
코드는 전체적으로 잘 작성해주셨습니다 ㅎㅎ 하지만, 클래스가 너무 많은 역할을 하고 있는 것 같아요. 조금 더 분리해보면 좋을 것 같아서 관련해서 코멘트 드렸는데 확인 부탁드립니다....!

감사합니다 : )

dongha-byun committed:
Refector: 문자열 계산기 코드 리펙터링
- 문자열 검증 수행로직 class 분리
- 계산식 parsing 로직 class 분리
- class 분리에 따른 테스트코드 추가 작성

chae-yh commented on src/main/java/calculator/utils/StringCalculateParser.java:
```diff
@@ -0,0 +1,23 @@
+    public static final int CUSTOM_DELIMITER_INDEX = 1;
+    public static final int CUSTOM_CALCULATE_INDEX = 2;
+    public static final String CUSTOM_CALCULATOR_REGEX = "//(.)\n(.*)";
+    public static final Pattern CUSTOM_DELIMITER_STRING_COMPILE_PATTERN = Pattern.compile(CUSTOM_CALCULATOR_REGEX);
```
피드백 반영 👍

chae-yh commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,36 @@
+
+public class StringCalculator {
+
+    public static final int NULL_OR_EMPTY_RETURN_VALUE = 0;
```
해당 상수를 사용할 경우 StringCalculator.NULL_OR_EMPTY_RETURN_VALUE 로 사용하게 될텐데 너무 추상적인 표현인 것 같아요. 단순하게 ZERO(StringCalculator.ZERO)는 어떠세요...?

chae-yh commented on src/main/java/calculator/utils/CalculatorValidator.java:
```diff
@@ -0,0 +1,22 @@
+
+import java.util.regex.Pattern;
+
+public class CalculatorValidator {
```
역할을 잘 분리해주셨지만 이렇게 데이터 검증을 한곳에 몰아두지 않고 원시값 포장 / 일급컬렉션을 만들어주고 해당 클래스에서 값 검증을 해주면 어떨까요...?

chae-yh replied:
여기서의 의도는 예를 들어, `validateOnlyNumber` 같은 경우는 Number(혹은 Positive 정도...?)라는 원시값 포장 객체 생성 시 validation 체크가 가능하지 않을까.. 하는 점이었는데 이 부분은 어떻게 생각하시는지 궁금합니다...!

단지 검증이라는 이유로 하나의 클래스에 모아둔다면, 지금은 기능이 간단한 편이라 문제가 없겠지만 복잡한 기능이라면 validation만으로도 몇백, 몇천 라인의 클래스가 만들어질 것 같아요...!

chae-yh commented on src/test/java/calculator/utils/CalculatorValidatorTest.java:
```diff
@@ -0,0 +1,35 @@
+    @Test
+    @DisplayName("빈 문자열인지 확인한다.")
+    void checkInputIsEmpty(){
+        assertThat(CalculatorValidator.isEmpty("")).isTrue();
```
이렇게 assertThat이 나열되면 assertAll 을 써봐도 좋을 것 같습니다...!

chae-yh commented on src/test/java/calculator/utils/CalculatorValidatorTest.java:
```diff
@@ -0,0 +1,35 @@
+
+import static org.assertj.core.api.Assertions.*;
+
+public class CalculatorValidatorTest {
```
분리한 클래스에 대해 테스트 작성해주신 것 좋습니다 👍

chae-yh reviewed (changes requested):
동하님 안녕하세요. 피드백 반영 잘해주셨습니다 ㅎㅎ 👍 
다만 머지하기에는 조금 더 살펴보면 좋을만한 내용들이 보여 코멘트 드리니 한 번 더 확인 부탁드립니다...!

감사합니다 : )

dongha-byun committed:
Test: 테스트 코드 리펙터링

dongha-byun committed:
Refactor: 문자열 계산기 코드 리펙터링
- return 0 상수명 변경 (NULL_OR_EMPTY_RETURN_VALUE -> ZERO)

dongha-byun committed:
Test: 문자열 계산시 테스트코드 리펙터링
- 계산식 원시값 포장 처리로 인한 테스트코드 변경(String -> CalculateString)
- 계산식 검증 위치 변경(CalculatorValidator -> CalculateString)

dongha-byun committed:
Refactor: 문자열 계산기 프러덕션 코드 리펙터링
- 원시값 포장(String -> CalculateString)
- 계산식 검증 주체 변경(CalculatorValidator -> CalculateString)

chae-yh commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,36 @@
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
```
미사용 import 는 삭제 부탁드립니다...!

chae-yh commented on src/main/java/calculator/domain/CalculateString.java:
```diff
@@ -0,0 +1,26 @@
+public class CalculateString {
+    private String expression;
+
+    private CalculateString(){}
```
코딩 컨벤션이 적용되어 있지 않은 것 같네요. 확인 부탁드립니다 : )
해당 내용과 별개로, 객체 생성을 막으려고 하신 것 같은데, 좋습니다 👍

chae-yh commented on src/main/java/calculator/domain/CalculateString.java:
```diff
@@ -0,0 +1,26 @@
+package calculator.domain;
+
+public class CalculateString {
+    private String expression;
```
음.. 이 클래스의 역할은 무엇일까요...? 원시값을 포장하려는 의도는 좋습니다만, 해당 클래스가 특별한 역할을 하는 것 같지는 않습니다. 입력값에 대한 validation 체크 정도면 StringCalculator에서 해주어도 좋을 것 같아요

chae-yh commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,63 @@
+    @DisplayName("빈 문자열은 0으로 반환한다.")
+    void emptyStringIsZeroTest() {
+
+        assertThat(StringCalculator.addOperation(CalculateString.create(""))).isEqualTo(0);
```
```suggestion
        assertThat(StringCalculator.addOperation(CalculateString.create(""))).isZero();
```
도 가능할 것 같습니다 : )

chae-yh commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,63 @@
+
+    @ParameterizedTest
+    @ValueSource(strings = {"-1:2:3", "5,10:-4", "10000,-10", "-9"})
+    @DisplayName("음수가 나오는 경우, 예외를 발생시킨다.")
```
이 예외가 CalculatorValidator에서 `validateOnlyNumber` 동작 시 숫자인지 판단하는데서 exception을 발생시키네요. 음수는 숫자가 아닐까요...?

chae-yh commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,36 @@
+import static calculator.utils.CalculatorValidator.*;
+import static calculator.utils.StringCalculateParser.parseStringToArray;
+
+public class StringCalculator {
```
여기는 상태값 없이 static 메서드들만 있는데 생성자를 private로 막아 객체 생성을 막아 주어도 좋을 것 같아요..!

chae-yh commented on src/main/java/calculator/domain/CalculateString.java:
```diff
@@ -0,0 +1,26 @@
+
+    private CalculateString(){}
+
+    public static CalculateString create(String input) {
```
static factory 메서드를 의도하신 것 같군요 ㅎㅎ 좋습니다 👍 그런데 사용을 테스트에서만 하는 것 같네요. 테스트에서만 사용하는 메서드는 가능한 지양하는 편이 좋을 것 같습니다...!

chae-yh reviewed (approved):
동하님 안녕하세요...! 피드백 반영해주신 것 확인했습니다. 하지만 코드 내에서 조금 더 고민해보시면 좋을만한 내용들이 보여 코멘트 드리니 확인 부탁드립니다...!

step2는 여기서 merge하겠습니다. 다음 step부터 본격적인 로또 미션이네요. 파이팅 입니다 : )
